### DISCUSSION TITLE
Websockets + broadcasting

### DISCUSSION THREAD

**Discussion by berndmehnert** [rwf2/Rocket#2602] [Questions] [answered]
Hi there,

thank you very much for this project. I am writing a small chat program in rust (i am quite new to rust ;). I use Rocket together with bootstrap5.3 for this) 
Since this is only for home usage at the moment, I use the master branch, since there we have websockets at our hands ;). 
In the "upgrade" example of the project, we have a small example of how websockets can be used in Rocket. But if I see this correctly, there is no broadcasting used. When a client writes a message, then the echo is not transmitted to all other clients, as for example in the chat example using SSEs. 

My Question would be: How would you implement broadcasting? Somehow we should have the necessary information from every client which is connected and then send any incomming message back to all clients... Thank you for any help ;)

**Comment by robertwayne** [comment, marked as answer]
Generally the way you handle this is to keep your own collection of connections. You can shove it into Rockets shared state so you can access it from your handlers.

Then to broadcast, you simply iterate over the connections and send. This also allows you to selectively broadcast by filtering, etc.

This is how it is done under the hood in every (?) WS library that offers a built-in broadcast from what I've seen.

**Reply by berndmehnert** [reply]
@robertwayne Thank you, I looked up the "shared state" in the documentation and that should really help here ;).

**Reply by SergioBenitez** [reply]
Note that the chat example in the repository broadcasts by using a broadcasting channel in managed state.

**Reply by berndmehnert** [reply]
Wow, I had it before my eyes in the example and did not see it ;). If I hopefully see this correctly, I could use that same broadcast-queue scenario, or something similar, but this time in the /post endpoint expect a ws::Websocket object, I will test this. Thank you ;)
